app/controllers/application_controller.rb

```diff
@@ -11,4 +11,22 @@ def after_sign_in_path_for(user)
     pages_url
   end
 
+  private
+
+  def localize_from_page_id
+    page = Page.find_by(id: params[:page_id])
+    if page.present? && page.language.present? && page.language.code.present?
+      set_locale(page.language.code)
+    end
+  end
+
+  def set_locale(code)
+    begin
+      I18n.locale = code
+    rescue I18n::InvalidLocale
+      # by setting the +i18n.enforce_available_locales+ flag to true but
+      # catching the resulting error, it allows us to only set the locale
+      # if it's one explicitly registered under +i18n.available_locales+
+    end
```

Comment:
What I wanted was a way to set the locale if and only if we had a valid locale code, and otherwise fall back to the default locale. I then changed a setting to raise the error, and set this up to catch it, giving me the desired behavior. I think we'd like to know about the error, but I don't think the user does. Open to changing it, but that was my intention.